### PR TITLE
If we can't statically prove 32-bit indexing OK, only add guard if hint exists

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106004

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -2171,11 +2171,17 @@ def reduction_hint(node):
     def can_use_32bit_indexing(numel: sympy.Expr, buffers: Iterable[ir.Buffer]) -> bool:
         int_max = torch.iinfo(torch.int32).max
         size_hint = V.graph.sizevars.size_hint
-        if size_hint(numel) > int_max:
-            return False
+        has_hint = V.graph.sizevars.shape_env.has_hint
+
+        if not V.graph.sizevars.is_expr_static_and_true(numel <= int_max):
+            if not has_hint(numel) or size_hint(numel) > int_max:
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:2177]: Can you please factor out this condition into its own function and use here and in the `any`.
A small comment like
```
# if we can't prove that its less than 2**32 and we have that it is an unbacked int or its current value is larger than 2**32, then we can't  use 32 bit arithmetic.
```

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -3097,6 +3097,17 @@ def size_hint(self, expr: "sympy.Expr"):
...
+    # NB: keep in sync with size_hint
+    @lru_cache(256)
+    def has_hint(self, expr: "sympy.Expr"):
+        result_expr = safe_expand(expr).xreplace(self.var_to_val)
+        if len(result_expr.free_symbols) != 0:
+            r = self._maybe_evaluate_static(result_expr)
+            if r is not None:
+                return True
+            return False
+        return True
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:3109]:
```suggestion
    return len(result_expr.free_symbols) == 0 or self._maybe_evaluate_static(result_expr) is not None
```

**Review by lezcano — APPROVED** [review]
Two small readability comments, otherwise lgtm

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**